src/InternPool.zig

```diff
@@ -5315,7 +5315,7 @@ pub fn get(ip: *InternPool, gpa: Allocator, key: Key) Allocator.Error!Index {
 
             try ip.extra.ensureUnusedCapacity(
                 gpa,
-                @typeInfo(Tag.Aggregate).Struct.fields.len + @as(usize, @intCast(len_including_sentinel)),
+                @typeInfo(Tag.Aggregate).Struct.fields.len + @as(usize, @intCast(len_including_sentinel + 1)),
```

Comment:
The CI run for Linux ran into a buffer overflow panic when initializing a sentinel-terminated array for reasons seemingly unrelated to the `@typeInfo` changes.
https://github.com/ziglang/zig/actions/runs/7439194713/job/20238792974

I based this fix on https://github.com/ziglang/zig/blob/804cee3b93cb7084c16ee61d3bcb57f7d3c9f0bc/src/InternPool.zig#L5274

I'm not at all familiar with compiler internals so I could be completely off base, but from skimming through code from `zirArrayInit` to here it appears to me that the problem is that `aggregate.storage.elems` includes the sentinel

```zig
const final_len = try sema.usizeCast(block, src, array_ty.arrayLenIncludingSentinel(mod));
const resolved_args = try gpa.alloc(Air.Inst.Ref, final_len);
// ...
const elem_vals = try sema.arena.alloc(InternPool.Index, resolved_args.len);
// ...
const arr_val = try mod.intern(.{ .aggregate = .{
    .ty = array_ty.toIntern(),
    .storage = .{ .elems = elem_vals },
} });
```

which makes these two operations require an available capacity of `len_including_sentinel + 1`

```zig
ip.extra.appendSliceAssumeCapacity(@ptrCast(aggregate.storage.elems));
if (sentinel != .none) ip.extra.appendAssumeCapacity(@intFromEnum(sentinel));
```

Judging by the related (?) fix in 5fa260ba065f28632af133d07441d8a611fba15b it's possible that `elems` should be sliced to exclude the sentinel

```diff
-ip.extra.appendSliceAssumeCapacity(@ptrCast(aggregate.storage.elems));
+ip.extra.appendSliceAssumeCapacity(@ptrCast(aggregate.storage.elems[0..@intCast(len)]));
 if (sentinel != .none) ip.extra.appendAssumeCapacity(@intFromEnum(sentinel));
```

but I have no idea what the consequences of doing so would be so I decided not to touch it.

Reply:
Thanks for the notes